### PR TITLE
Bluetooth: Mesh: Generic OnOff & Level Models demo

### PR DESCRIPTION
**Pull request by vikrant8052**
This is a simple application demonstrating a Bluetooth mesh node in
which Root element has following models

 - Generic OnOff Server
 - Generic OnOff Client
 - Generic Level Server
 - Generic Level Client

And Secondary element has following models

 - Generic OnOff Server
 - Generic OnOff Client

Signed-off-by: Vikrant More vikrant8051@gmail.com

### CODE REVIEW

File: samples/boards/nrf52/mesh/onoff_level_app/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+sample:
+  name: Bluetooth Mesh
+tests:
+  test:
+    harness: bluetooth
+    platform_whitelist: bbc_microbit nrf51_blenano qemu_x86
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/sample.yaml]: This sample is specifically for nRF52, so the above whitelist looks wrong to me. It should only contain one or more nRF52 targets. The same applies for the various *.conf files for microbit, etc. There should only be a single prj.conf that targets the nRF52.

File: samples/boards/nrf52/mesh/onoff_level_app/src/main.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		light_state_current.power = LEVEL_100;
+	}
+
+	printk("\n\rled_on_off_status -> %d, led_level_status -> %d\n\r", light_state_current.OnOff, light_state_current.power);
+
+	gpio_init();
+
+	update_light_state();
+
+	printk("Initializing...\n\r");
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/main.c]: Use only \n for the newlines. You're now using \n in some places and \n\r in others (like here).

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/DEVICE_composition.c [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+const struct bt_mesh_comp comp = {
+	.cid = CID_INTEL,
+	.elem = elements,
+	.elem_count = ARRAY_SIZE(elements),
+};
+
+void gen_onoff_get(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf)
+{
+	doer(model, ctx, buf, 0x8201);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/DEVICE_composition.c]: You really need to come up with a better name than "doer". `process_message()` perhaps?

File: samples/boards/nrf52/mesh/onoff_level_app/src/board.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/* board.h - Board-specific hooks */
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/board.h]: This file seems unnecessary since you're only targetting the nRF52 (i.e. a single board). I know it's there in other samples like samples/bluetooth/mesh, however those are designed to be adapted potentially on multiple different boards.

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/DEVICE_composition.c [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	{ BT_MESH_MODEL_OP_2(0x82, 0x03), 2, gen_onoff_set_unack },
+	BT_MESH_MODEL_OP_END,
+};
+
+static const struct bt_mesh_model_op gen_onoff_cli_op[] = {
+	{ BT_MESH_MODEL_OP_2(0x82, 0x04), 1, gen_onoff_cli_status },
+	BT_MESH_MODEL_OP_END,
+};
+
+static const struct bt_mesh_model_op gen_level_srv_op[] = {
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/DEVICE_composition.c]: Move this lower down in the file so that you can make all the handler functions static.

File: samples/boards/nrf52/mesh/onoff_level_app/src/microbit.c [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+void board_prov_complete(void)
+{
+	struct mb_display *disp = mb_display_get();
+	struct mb_image arrow = MB_IMAGE({ 0, 1, 0, 1, 0 },
+					 { 0, 1, 0, 1, 0 },
+					 { 0, 0, 0, 0, 0 },
+					 { 1, 0, 0, 0, 1 },
+					 { 0, 1, 1, 1, 0 });
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/microbit.c]: This image is a smiley face and not an arrow, so better fix the name or the image :) I realize this is probably a copy-paste from somewhere else but let's not replicate the error (instead, please submit a separate fix to the original one).

File: samples/boards/nrf52/mesh/onoff_level_app/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(NONE)
+
+target_link_libraries(app subsys__bluetooth)
+
+target_sources(app PRIVATE
+	       src/main.c
+	       src/common.c
+	       src/mesh/ble_mesh.c
+               src/mesh/DEVICE_composition.c
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/CMakeLists.txt]: Why the upper case `DEVICE`? Please be consistent (also, file names are generally lower case in Zephyr). Also, it seems the indentation is off here compared to the previous lines.

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c [hunk 8]
```diff
@@ -0,0 +1,182 @@
+#include "common.h"
+
+void doer(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf, uint16_t opcode)
+{
+	uint8_t tid, tmp8;
+	int16_t tmp16;
+	int err = 0, tmp32;
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c]: If you're going to call it tmp32, better make it s32_t or u32_t.

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+	struct net_buf_simple *msg;
+
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c]: Unnecessary empty line

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+	struct server_state *state_ptr = model->user_data;
+
+	switch (opcode) {
+
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c]: Unnecessary empty line. In fact, it almost looks like you took a different style in this file with lots more empty lines than your other code. Why is that?

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+		break;
+
+	case 0x8203:	/* GEN_ONOFF_SRV_UNACK */
+
+		msg = model->pub->msg;
+
+		tmp8 = net_buf_simple_pull_u8(buf);
+
+		state_ptr->current = tmp8;
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c]: Why is tmp8 needed here? Why not assign to state_ptr->current directly?

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+void doer(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx, struct net_buf_simple *buf, uint16_t opcode)
+{
+	uint8_t tid, tmp8;
+	int16_t tmp16;
+	int err = 0, tmp32;
+
+	struct net_buf_simple *msg;
+
+	struct server_state *state_ptr = model->user_data;
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/doer.c]: We don't generally put type information in variable names, so please remove the `_ptr` part.

**Review by jhedberg — CHANGES_REQUESTED** [review]
There are many more issues I'd like to comment on, but ran out of time for now. I'll take another look once you've sent the next revision.

**Comment by vikrant8051** [comment]
@jhedberg I hope now everything is Okay !! 

There is some error regarding email address. Is that means I can't merge this PR with my personnel email ID ? 
**author email (vikrant8051 <vikrant8051@gmail.com>) does not follow the syntax: First Last <email>.**

How to edti previous commit messages ? 
**failure: commit message error on range: origin/master..HEAD**

**Comment by jhedberg** [comment]
> There is some error regarding email address. Is that means I can't merge this PR with my personnel email ID ? author email (vikrant8051 vikrant8051@gmail.com) does not follow the syntax: First Last .

I don't understand what that has to do with your email. The name that precedes the email address is completely independent of the email address. The simplest would be if you just add the following to your `~/.gitconfig`:

```
[user]
	name = Vikrant More
	email = vikrant8051@gmail.com
```

That will then generate the right author information, and the right signed-off-by line when you pass `-s` to git commit.

I also recommend you get familiar with `git rebase -i`

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/process_message.c [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+		     struct net_buf_simple *buf, uint16_t opcode)
+{
+	uint8_t tid, tmp8;
+	int16_t tmp16;
+	s32_t err = 0, tmp32;
+	struct net_buf_simple *msg;
+	struct server_state *state = model->user_data;
+
+	switch (opcode) {
+	case 0x8201:	/* GEN_ONOFF_SRV_GET */
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/process_message.c]: I took a closer look at this function now (previously named "doer") why are you de-multiplex the message handlers into this single function, when the various case branches have essentially nothing in common (except perhaps some function variables). I'd either split this up to one function per message, or even better just merge the code back to the message handlers that call this function. Why was there a need to do this outside of the message handlers in the first place? (if there is a well justified need I'd still split this up into one function per message, and then you also solve the trouble you had with naming, since you can name the functions based on the message that they handle).

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/publisher.c [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+		bt_mesh_model_msg_init(root_models[5].pub->msg,
+				       BT_MESH_MODEL_OP_2(0x82, 0x0A));
+		net_buf_simple_add_le32(root_models[5].pub->msg, -100);
+		net_buf_simple_add_u8(root_models[5].pub->msg, tid_level++);
+		err = bt_mesh_model_publish(&root_models[5]);
+#endif
+	}
+
+	if (err) {
+		printk("bt_mesh_model_publish: err: %d\n\r", err);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/publisher.c]: still \r here

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/process_message.c [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+			bt_mesh_model_msg_init(msg,
+					       BT_MESH_MODEL_OP_2(0x82, 0x08));
+
+			net_buf_simple_add_le16(msg, state->current);
+
+			err = bt_mesh_model_publish(model);
+
+			if (err) {
+				printk("bt_mesh_model_publish err %d", err);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/process_message.c]: missing \n

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/process_message.c [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+		if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
+			bt_mesh_model_msg_init(msg,
+					       BT_MESH_MODEL_OP_2(0x82, 0x04));
+			net_buf_simple_add_u8(msg, state->current);
+
+			err = bt_mesh_model_publish(model);
+
+			if (err) {
+				printk("bt_mesh_model_publish err %d", err);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/process_message.c]: missing \n

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/ble_mesh.c [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+	if (err) {
+		printk("Bluetooth init failed (err %d", err);
+		return;
+	}
+
+	printk("Bluetooth initialized");
+
+	/* Use identity address as device UUID */
+	if (bt_le_oob_get_local(&oob)) {
+		printk("Identity Address unavailable");
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/ble_mesh.c]: missing \n

File: samples/boards/nrf52/mesh/onoff_level_app/src/mesh/ble_mesh.c [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+	.complete = prov_complete,
+	.reset = prov_reset,
+};
+
+void bt_ready(int err)
+{
+	struct bt_le_oob oob;
+
+	if (err) {
+		printk("Bluetooth init failed (err %d", err);
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/mesh/ble_mesh.c]: missing ) and \n

File: samples/boards/nrf52/mesh/onoff_level_app/src/main.c [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+		     &light_state_current,
+		     sizeof(struct light_state_t)) < 0) {
+
+		printk("\n\rentry for NVS_LED_STATE_ID is not found on NVS\n");
+
+		light_state_current.OnOff = 1;
+		light_state_current.power = LEVEL_100;
+	}
+
+	printk("\n\rled_on_off_status -> %d, led_level_status -> %d\n",
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_app/src/main.c]: still \r here in the beginning. Actually why do you start the format string with a new line? Isn't the one at the end sufficient?

File: samples/boards/nrf52/mesh/onoff-app/prj.conf [hunk 20]
```diff
@@ -2,6 +2,14 @@
 CONFIG_MAIN_STACK_SIZE=512
 CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
 
+CONFIG_BT_SETTINGS=y
+CONFIG_FLASH=y
+CONFIG_FLASH_PAGE_LAYOUT=y
+CONFIG_FLASH_MAP=y
+CONFIG_FCB=y
+CONFIG_SETTINGS=y
+CONFIG_SETTINGS_FCB=y
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff-app/prj.conf]: Please split this into a separate PR (that's what I intended when I asked for this addition). It can then probably be merged must faster than this PR, since I don't see anything wrong with the patch. If any of this feels like a lot of work, please get familiar with `git rebase -i`. That will get the work done in a matter of seconds, or a minute at most.

**Comment by vikrant8051** [comment]
@jhedberg I make suggested changes. Use ~/.gitconfig to change name. Hope will not face identity error. I will work on Commit message error tomorrow.  Thank You !!
